### ISSUE TITLE
anna -> adam support?

### ISSUE THREAD

**Issue by riemers** [laetificat/haanna#1]
Hi there fellow leidenaar :)

I bought some plugwise equipment and it works pretty nice. Then noticed there was no component for it :( then i saw your hanna component and i was 'hey this might work'.

Only to find out that adam is a bit different since it allows for more zones (uses toms for radiators, lisa's for thermostat screens and floors for floor heating)

Is there anyway we can incooperate this too? haadam or haaplugheat? Don't mind to show my face since your in the same city 👍

**Comment by riemers** [comment]
If you only have 1 lisa, then this already works the same:

```
location_id = root.find("appliance[type='zone_thermostat']/location").attrib['id']
#location_id = root.find("appliance[type='thermostat']/location").attrib['id']
```

Just change thermostat to zone_thermostat. However this will bork i presume when adding multiple.

As reference, there are more types you can get out of it:

```
thermostatic_radiator_valve (tom)
zone_thermostat (lisa)
heater_central (boiler)
gateway (Adam) (general gateway information)
open_therm_gateway (OpenTherm Gateway)
```

Noticed you had a ping for a 404 seems it has this one http://192.168.1.38/cache/gateways;ping

**Comment by riemers** [comment]
![image](https://user-images.githubusercontent.com/682041/53291723-1eb1f500-37b8-11e9-8bf6-8a96ed143bfb.png)

This is from my adam. Only missing is a pulldown to set away/home/etc. Presume you can set that for all too in some way.

**Comment by frankhommers** [comment]
This would be great to have this with more than one Plugwise Lisa!

**Comment by laetificat** [comment]
Hello, thank you for your interest in this project. I've been very busy lately so sorry for the late response (the recent activity actually triggered me to check up on this project again).

Implementing multiple devices is something I can take a look at, at least for the lisa's if I can find them.
The ping endpoint is a great tip and I will implement that for sure.

**Comment by riemers** [comment]
Well since you live in the same city (i presume) i can always help out with the hardware for testing if that would help you out.

**Comment by riemers** [comment]
No rush yet, but if you have some spare time i can always drop by with an adam and a lisa. I should really go get myself multiple lisas and valves so i can get the data out and see what is different. In all fairness i think that this component might even support then both making the anna name maybe not the right one. But that is the least of the issues. Did see updates on the pypi, all little things help.

@CoMPaTech unless you have some time, i see you live around utrecht.. i sometimes visit nieuwegein, is around the corner.

**Comment by laetificat** [comment]
@riemers I do not have much spare time now or in the near future, I'll try my best to be more active on this project but life tend to get in the way sometimes.
The API is not set up to handle multiple devices and zones, this plays a huge part in the Adam software. It is possible to refactor the code into supporting the Adam and I think that is the correct way to move this project forward to but it will be a huge task. Regarding the name, I think it should be more generic like Plugwise Python API or something once Adam is supported.

**Comment by riemers** [comment]
Agreed, i'll get some devices more for in my house to play around. You could also just add compatech to your repo for the time being since he helped too. In case your too busy, this way we all try to help out. (i got 2 kids, i know how busy stuff can be)

**Comment by laetificat** [comment]
I added compatech to the repo, thanks for the tip!

**Comment by riemers** [comment]
@CoMPaTech this one is still open, i'll leave it here for the future. My valves are not ok for plugwise floor heating so i first need to get those replaced (some weird dimensions)

**Comment by riemers** [comment]
Took longer then expected, but i now have a lisa/adam/floor and a tom. So i think pretty much the whole stack sort of speak. @CoMPaTech i know you would want to take a look at it, is there any data specific you want? Outputs of certain things? Always good to have.

**Comment by ghost** [comment]
I installed Adam, Lisa and a Tom last weekend and would love to have the thermostat in HA! I'll also be glad to help with testing and supplying information, and if I can find time will look into help in development of the plugin. I don't really have experience with Python yet, but it might be a good way to learn something about it :)

**Comment by frankhommers** [comment]
Yeah, I also have Adam HA in production for quite a while now. It manages multiple zones. Only thing I am missing is nice Home Assistant integration ;-)

**Comment by CoMPaTech** [comment]
Is there any XML exposed like our Smile's (the boiler device behind Anna) have? i.e. `192.168.x.y/core/domain_objects` and accompanying `/core/rules` `/core/appliances' and maybe something like `/#zones/...`

The way we control Anna, from @laetificat 's findings is by reading the XML and posting accordingly what change we need from inspecting what the browser interfaces. I.e. reading is just interpreting said XML, writing is compromised by inspecting the network traffic from the browser.

**Comment by XMaarten** [comment]
@CoMPaTech I have a Adam as well, and I attached a couple of XML files for you. My living room uses a floor radiator valve, with a separate zone thermostat because the valve itself is in a different room, and three rooms with conventional radiators. 

Below the XML files:
[rules.txt](https://github.com/laetificat/haanna/files/3908440/rules.txt)
[appliances.txt](https://github.com/laetificat/haanna/files/3908441/appliances.txt)
[domain_objects.txt](https://github.com/laetificat/haanna/files/3908442/domain_objects.txt)
[locations.txt](https://github.com/laetificat/haanna/files/3908443/locations.txt)

**Comment by XMaarten** [comment]
A single thermostat used to work in a previous version of HomeAssistant, but since the 0.102 update this fails with the following error:
```
ERROR (MainThread) [homeassistant.components.climate] plugwise: Error on device update!
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/homeassistant/helpers/entity_platform.py", line 292, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/local/lib/python3.7/dist-packages/homeassistant/helpers/entity.py", line 450, in async_device_update
    await self.hass.async_add_executor_job(self.update)
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.7/dist-packages/homeassistant/components/plugwise/climate.py", line 267, in update
    self._preset_mode = self._api.get_current_preset(self._domain_objects)
  File "/home/homeassistant/.homeassistant/deps/lib/python3.7/site-packages/haanna/haanna.py", line 362, in get_current_preset
    return root.find(locator).text
AttributeError: 'NoneType' object has no attribute 'text'
```

**Comment by riemers** [comment]
@CoMPaTech @bouwew a new year new chances.. i would really like to see Adam support. What would be required to make a start, do you need details? The devices? Do you need a adam to test with etc?

Wait, i can do better. Found a shop that is willing to donate a adam, lisa, floor, tom for support in HA. You can always make a PR later to HA and mention the shop (since it is nice gesture) Only question would be, does it have extra benefit for either @CoMPaTech or @bouwew since they have a anna now. 🙍‍♂

**Comment by bouwew** [comment]
@XMaarten @riemers 
I quickly looked at the XML-data provided. I see quite a few similarities with the data in the Anna XML-files. It shouldn't be too difficult to create something.
My main question is, what would you like to see in HA as a start? Meaning what data/values would you like to see? Next, what kind of programming would you want to be able to do?
Can you share picture from the App that you are using?

**Comment by bouwew** [comment]
@riemers being practical, I have floor-heating downstairs and radiators upstairs.
In my house there is an Intergas CV-ketel that is the heat-source.

I could put up a Lisa and a Tom in my bathroom. And also an Adam is needed, I see in the docs that the Anna can be connected to the Adam as well. So, yes, I would be interested to experiment with these parts and try to integrate them into HA.

**Comment by riemers** [comment]
In that case, drop me a mail at riemers@binkey.nl so we can talk, or just on discord Riemers
#4426

**Comment by bouwew** [comment]
@XMaarten can you please capture the XML at /core/direct_objects as well? If that path is on your Adam?

**Comment by XMaarten** [comment]
@bouwew 
What I would like to do is to see and control the temperature of the four different rooms I have. 
I attached the direct_objects and a screenshot of the Plugwise app.

[direct_objects.txt](https://github.com/laetificat/haanna/files/4049700/direct_objects.txt)
![plugwise](https://user-images.githubusercontent.com/58396778/72209133-6ebaa880-34ab-11ea-9fa0-ed3c583951c0.jpeg)

**Comment by bouwew** [comment]
@XMaarten
One more question: which hardware items are in your system and how many?

**Comment by XMaarten** [comment]
@bouwew The temperature of the living room is measured by the thermostat, and the radiator valve (Floor) controls the floor warming. The valve is not in the living room itself so the temperature measured of the floor warming can not be used, I added a separate thermostat for measuring the temperature in the living room. Besides the living room I have two sleeping rooms and a bathroom controlled by their own radiator valve (Tom), which measure their own temperature in the room, so four rooms in total.

**Comment by riemers** [comment]
That’s why we have a Lisa too. That measures the temperature and allows control of a floor.

________________________________
Van: XMaarten <notifications@github.com>
Verzonden: Sunday, January 12, 2020 1:24:16 PM
Aan: laetificat/haanna <haanna@noreply.github.com>
CC: Erik-jan Riemers <riemers@binkey.nl>; Mention <mention@noreply.github.com>
Onderwerp: Re: [laetificat/haanna] anna -> adam support? (#1)


@bouwew<https://github.com/bouwew> The temperature of the living room is measured by the thermostat, and the radiator valve (Floor) controls the floor warming. The valve is not in the living room itself so the temperature measured of the floor warming can not be used, I added a separate thermostat for measuring the temperature in the living room. Besides the living room I have two sleeping rooms and a bathroom controlled by their own radiator valve (Tom), which measure their own temperature in the room, so four rooms in total.

**Comment by bouwew** [comment]
@XMaarten I need you to be more specific to be able to make sense of the data in the XML-files.
1x Adam HA I would assume?
1x thermostat - Plugwise / other brand, what type (OpenTherm or on/off), connected to the Adam?
1x Floor
3x Tom
Anything else?

**Comment by XMaarten** [comment]
I am sorry.
1x Adam HA, connected to the boiler with on/off, I need to replace my boiler to a OpenTherm one.
1x Floor (radiator valve for the living room)
1x Lisa (thermostat / sensor for the living room)
3x Tom (for sleepingrooms and bathroom)

Lisa is not connected to anything else, it is only used as temperature sensor for the Adam, to measure the temperature in the living room.

**Comment by bouwew** [comment]
Great, thanks!

**Comment by bouwew** [comment]
I had a detailed look at the XML-data, my first impression is that there will be several climate-entities, 4 in XMaartens system, instead of 1 for the Anna in my system, each with the possibility to set presets and activate a weekly schedule. Is it possible to create your own weekschedule per "thermostat"? And is it possible to create more than one schedule per "thermostat"? Or can there be only one weekschedule per "thermostat"?

**Comment by XMaarten** [comment]
Yeah, that is possible, you can give each room its own schedule.

**Comment by bouwew** [comment]
And what about my question about having more than one schedule per room? Possible? Or not?

**Comment by XMaarten** [comment]
You can define schedules in the plugwise app and specify what schedule is applicable to which room. So yes, you can define multiple schedules.

**Comment by frankhommers** [comment]
I can help with test data as well. I have 4 rooms defined here 3 lisa's, 4 plugs and 1 adam-ha.
Tell me what you need ;-)

**Comment by riemers** [comment]
@bouwew will be sponsered with some hardware from https://hashop.nl so he has all the devices to play about. Does not hurt to give the xml's too so he has some more different ones too. He has some personal items on his agenda, when that has settled he can take a swing at the code part.

**Comment by frankhommers** [comment]
Ah I'll wait from specific requests from @bouwew then.

**Comment by bouwew** [comment]
@XMaarten I'm working with your files, I think you have defined 2 weekschedules for K***'s bedroom, correct? Both with the same name: K***.
Would you mind if I set a rule to use different names for different schedules? That will make my (programming) life much easier :)

About the progress, this is the present output of my code:
```
{'95395fb15c814a1f8bba88363e4a5833': 'zone_thermostat', 'e4fe90f9b4c44887b09f8640e828eb3b': 'thermostatic_radiator_valve', '450d49ef2e8942f78c1242cdd8dfecd0': 'thermostatic_radiator_valve', 'bc9e18756ad04c3f9f35298cbe537c8e': 'thermostatic_radiator_valve', '77ad7c755e0b477ebb1dc1361252832b': 'thermostatic_radiator_valve', '412c73075c4040058118e65cde1f0be9': 'heater_central'}
95395fb15c814a1f8bba88363e4a5833 zone_thermostat
{'6d42c3e3fa384294ad76780120eb7a88': 'Living room'}
Living room
{'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}
{'Living room': True}
e4fe90f9b4c44887b09f8640e828eb3b thermostatic_radiator_valve
{'75ab287de4194550bec50477f46525be': 'Bedroom 1'}
Bedroom M******
{'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}
None
450d49ef2e8942f78c1242cdd8dfecd0 thermostatic_radiator_valve
{'fc8b90f96c76416a81fea1921e16827c': 'Bedroom 2'}
Bedroom K***
{'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}
{'K***1': False, 'K***2': True}
bc9e18756ad04c3f9f35298cbe537c8e thermostatic_radiator_valve
{'6d42c3e3fa384294ad76780120eb7a88': 'Living room'}
Living room
{'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}
{'Living room': True}
77ad7c755e0b477ebb1dc1361252832b thermostatic_radiator_valve
{'01bc059aef734090bfad1d01c9145ca0': 'Bathroom'}
Bathroom
{'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}
{'Bathroom': True}
412c73075c4040058118e65cde1f0be9 heater_central
{}
```

First I'm trying to find all the devices that are present in the system and that I'm interested in. I also collect the ID's which I can use to find linked information. 
You can see the Lisa, the 4 Tom's and the heating device. There are more but of no interest at the moment.

Next I'm trying to find the locations that you have defined in your home and their real names.
For instance I see Bedroom 1 and Bedroom M****** in the various files, these two are linked. I assume the name Bedroom 1 is created by the system itself and you have entered the name Bedroom M******. Is my assumption correct?

Next, I'm finding the presets for each location (I still need to add the status for each preset, active or not)

Finally, I'm finding the defined weekschedules for each location and their status active (True) or not (False). As you can see, I've renames the schedules named `K***` to `K***1` and `K***2`.

More to follow...

**Comment by XMaarten** [comment]
I guess I forgot to obfuscate some names from the XML, sorry. 

I have no objection against using unique names in the different schedules, and I don't care that much about the schedules anyway. Most important for me is the view the current and the set temperature, and adjust the set temperature as needed, I can always use the plugwise app to define the schedules. If you ask me, I would just leave them out.

**Comment by riemers** [comment]
I would ignore schedules, those you setup in a adam/anna. Sure it would be nice, but do not forget that both those devices have a learning pattern that they use to get the ideal temperature in a correct manor. It takes about 2 weeks for the system to work optimal. Those kind of logic is not present as of now in HA. Temperature up/down and to set the modes "vacation/away/sleeping" etc would be the first step.

**Comment by bouwew** [comment]
Ok, great :)

My goal is to create the same functionality as is now present in the Plugwise Anna code, as far as this is possible. And, if possible, support both Anna/Smile and Adam systems.
I just tested my latest code on my Anna, the info that I want to obtain is all there and no errors! Looks like I'm on the right track :)

@riemers, when comparing the Adam-XML-code to the Anna-Smile-XML-code I see a lot of similarities. So, what is possible on the Anna is possible on the Adam, is my working assumption.
I will find out when I actually install the Adam, etc. in my home.

For now, I will try to finish the "obtaining-data" part, meaning create the plugwise library and a matching climate.py file that the interested people can install as a custom_component.
Next, for that I will need to have the Adam installed, I will tackle the "programming"-part.

**Comment by riemers** [comment]
You can use HACS for a custom component then do all your work in the "dev" branch, people can select that branch then in hacs i believe to try it out easilly. Once working you can push it to master and eventually send a PR to HA. But i know that will take some time hence HACS in between. (although i don't mind downloading from somewhere too)

**Comment by bouwew** [comment]
@riemers I have a default weekschedule programmed in my Anna that I run continuously when I am home. Only when I'm away for a longer time, I switch to a preset.
When looking at the Adam-XML-code from XMaarten I can see that the Adam is basically providing the Anna-functionality for every location where there is an Anna, a Lisa or a Tom present. So don't worry about it :)

**Comment by bouwew** [comment]
I know HACS, I will use it when the component is mostly done. But first I will post the link to my github here, so the early adopters can test first. When the testing within this limited group is successful then I'll put it in HACS, like some of my other plugins that are in HACS now.

**Comment by bouwew** [comment]
@laetificat @CoMPaTech 
Please review my work on the new plugwise library: https://github.com/bouwew/plugwise
Am I taking the right approach with this output?

For illustration, this is the output I obtain from my Smile/Anna:
```
[['Living room', None, '20.00', '20.25', {'home': 20.0, 'vacation': 15.0, 'no_frost': 10.0, 'asleep': 18.0, 'away': 18.0}, ['Test', 'Weekschema'], 'Weekschema', 'Weekschema'], ['Controlled Device', '34.0', None, False, None, False, '6.0', '6.9']]
```
Please see the README.md on the github for an explanation of the data-items.

Please note: only the functions that obtain data are working, I have not done any work on setting stuff. This code is copied over from haanna.

**Comment by bouwew** [comment]
And this is the output from a set of Adam-User files:
```
[['Living room', '0.84', '20.00', '19.70', {'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}, ['Living room'], 'Living room', 'Living room'], ['Bedroom M***', '0.48', '21.00', '20.93', {'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}, [], None, None], ['Bedroom K***', '0.77', '20.00', '20.18', {'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}, ['K***', 'K***'], 'K***', 'K***'], [], ['Bathroom', '0.88', '17.00', '19.13', {'home': 20.0, 'asleep': 17.0, 'away': 15.0, 'vacation': 15.0, 'no_frost': 10.0}, ['Bathroom'], 'Bathroom', 'Bathroom'], ['Controlled Device', '80.0', None, False, None, None, None, '3.0']]
```
This user has, next to the Adam, a Lisa and four Toms. The Lisa and 1 Tom are in the living room, I think.

I've started to work on the custom_component files: `___init___.py`, `climate,py`, etc. It might be that I'll change the format when that is more convenient for the code inside these python files.

**Comment by CoMPaTech** [comment]
Added review in the code repo issues. With regard to multiple devices (i.e. two Toms and a Lisa) in the living we might need to figure if we change the proposed dict to just have a room name and all thermostat devices under that (also see review issue).

**Comment by bouwew** [comment]
@CoMPaTech 
My working assumption based on the information that I have (from the XML-data I received, and the descriptions and manuals on the Plugwise website), is that the Lisa is the master of the Tom's in the configured zone. The Tom's are the slave devices and follow the setpoints received from the master. I have assumed for now that I can ignore the data from the slave devices and only provide HA with data from the master devices (the Lisa for the Living Room and the Tom's in the Bedrooms and the Bathroom). In the future we could consider adding this data, if there is a need for it.

**Comment by bouwew** [comment]
@XMaarten can I contact you in another way then from here? Via Discord or some other platform?

**Comment by XMaarten** [comment]
Sure. I have Skype, e-mail, telegram and such, but not discord. I tried sending you (or someone with the same name) a message on Skype.

**Comment by bouwew** [comment]
Adam support is here!

With some help from https://hashop.nl I have been able to develop a Home Assistant custom_component for the Plugwise Adam HA.

Please go here for the files and configuration-info: https://github.com/bouwew/adam

Any questions? Please post in this topic.

**Comment by riemers** [comment]
A lot of testing and patience from bouwe. But it is good to see such progress. I will think a lot of people will benefit of this one in the future.

**Comment by CoMPaTech** [comment]
@riemers in terms of good news, I've made a working proof of concept on having climate + sensor and once this dust settles we should be able to make some traction on that as well making the other request (+ denies from HA on bouwew's previous PR) play along nicely :)

**Comment by riemers** [comment]
Nice, indeed allowing sensors to turn off
My climate (door is open) would be the next thing I would look intoz

**Comment by bouwew** [comment]
@riemers With sensors CoMPaTech means there will be additional sensors-entities in HA that are related to the climate-entities. For instance showing the water-temperature inside the heater or the battery-percentage as reported by a Tom or a Lisa.

**Comment by riemers** [comment]
Sounds great, before we know it we have the WAF factor at 100% (although we all know that is impossible)

**Comment by bouwew** [comment]
@frankhommers @RonFlame In case you've missed my original message, Adam support is now available, please see a few posts up.

**Comment by frankhommers** [comment]
Great! @bouwew !!
Did this already end up in HA itself?

**Comment by bouwew** [comment]
Not yet, but we have created a PR, for climate only. Once this one is in we will add the rest: binary_sensors, sensors, switches, and water_heater.

**Comment by bouwew** [comment]
@frankhommers @RonFlame and others,
Update: please find our latest version here: https://github.com/plugwise/plugwise-beta
Make sure to completely remove the previously used custom_component-files and the entry in configuration.yaml, then reboot HA.
Next, make sure to read the README.md at the new github! Then install the new custom_component, enjoy!

**Comment by CoMPaTech** [comment]
@frankhommers @RonFlame @riemers @XMaarten we've just released 0.2.0 of our custom_component (with an upstream HA-PR in progress). As per the release of HA-core 0.109 this wednesday (or if you already run HA-core beta 109) you can start using `plugwise-beta`.

Closing this issue. If you have any further comments or issues feel free to update
 - The (community post)[https://community.home-assistant.io/t/plugwise-smile-custom-component-beta/183560/3?u=compatech]
 - The (module)[https://github.com/plugwise/Plugwise-Smile] replacing this repository
 - The (custom_component)[https://github.com/plugwise/plugwise-beta]
